### PR TITLE
Tt ovs 2.5

### PR DESCRIPTION
**Pull request by yearsj**
Dp - the basic process code is completed.

### CODE REVIEW

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]
you should add some comment

**Review by zhanghaooo — APPROVED** [review]
LGTM

File: datapath/Modules.mk [hunk 1]
```diff
@@ -24,6 +24,7 @@ openvswitch_sources = \
 	flow.c \
 	flow_netlink.c \
 	flow_table.c \
+    tt.c \
```
Comment (chenwhql) [inline datapath/Modules.mk:27]: consistent with the original style
image

File: datapath/tt.c [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    return pop_tt(skb);
+}
+
+
+//分配一个新的tt_table_item
+struct tt_table_item *tt_table_item_alloc(void) {
+	struct tt_table_item *item;
+
+	item = kmalloc(sizeof(*item), GFP_KERNEL);
+    if (!item) {
```
Comment (chenwhql) [inline datapath/tt.c:132]: please align code

File: datapath/tt.c [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+// tt_table重新分配空间
+static struct tt_table* tt_table_realloc(struct tt_table *old, int size) {
+	struct tt_table *new;
+
+	new = tt_table_alloc(size);
+	if (!new) {
+		printk(KERN_ALERT "DEBUG: tt_table realloc fail!  %s %d \n", __FUNCTION__, __LINE__);
+        return NULL;
+    }
```
Comment (chenwhql) [inline datapath/tt.c:178]: same as last one

File: datapath/tt.c [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+	if (old) {
+		int i;
+
```
Comment (chenwhql) [inline datapath/tt.c:182]: is it better to define local variable at head into function?

File: datapath/tt.c [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+// 删除指定flow_id的表项
+// 错误返回NULL
+struct tt_table* tt_table_delete_item(struct tt_table* cur_tt_table, __be16 flow_id) {
+	struct tt_table_item* tt_item;
+	if (!cur_tt_table || flow_id > cur_tt_table->max) {
+        printk(KERN_ALERT "DEBUG: flow_id to big! tt_table_item delete fail!  %s %d \n", __FUNCTION__, __LINE__);
+        return NULL;
+    }
+
```
Comment (chenwhql) [inline datapath/tt.c:219]: align

File: datapath/tt.c [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+	tt_item = ovsl_dereference(cur_tt_table->tt_items[flow_id - 1]);
+	if (tt_item) {
+		RCU_INIT_POINTER(cur_tt_table->tt_items[flow_id - 1], NULL);
+		cur_tt_table->count--;
+        call_rcu(&tt_item->rcu, rcu_free_tt_table_item);
+	}
```
Comment (chenwhql) [inline datapath/tt.c:225]: align

File: datapath/tt.c [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+	//必要时缩小flow_id
+	if (cur_tt_table->max >= (TT_TABLE_SIZE_MIN * 2) &&
+		cur_tt_table->count <= (cur_tt_table->max / 3)) {
+		struct tt_table* res_tt_table;
+		res_tt_table = tt_table_realloc(cur_tt_table, cur_tt_table->max / 2);
+		if (!res_tt_table) {
+            printk(KERN_ALERT "DEBUG: tt_table_item delete fail!  %s %d \n", __FUNCTION__, __LINE__);
```
Comment (chenwhql) [inline datapath/tt.c:233]: align

**Review by chenwhql — COMMENTED** [review]
please check the code style and format, maybe the code editing envrionment should be reconfig.

**Pull request merged**